### PR TITLE
[flang][OpenMP] Decompose compound constructs, do recursive lowering

### PR DESCRIPTION
**Pull request by kparzysz**
A compound construct with a list of clauses is broken up into individual leaf/composite constructs. Each such construct has the list of clauses that apply to it based on the OpenMP spec.

Each lowering function (i.e. a function that generates MLIR ops) is now responsible for generating its body as described below.

Functions that receive AST nodes extract the construct, and the clauses from the node. They then create a work queue consisting of individual constructs, and invoke a common dispatch function to process (lower) the queue.

The dispatch function examines the current position in the queue, and invokes the appropriate lowering function. Each lowering function receives the queue as well, and once it needs to generate its body, it either invokes the dispatch function on the rest of the queue (if any), or processes nested evaluations if the work queue is at the end.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Is it possible to add tests for this?

Can we move the decomposition logic into a separate file?

**Comment by kparzysz** [comment]
> Is it possible to add tests for this?
> 
> Can we move the decomposition logic into a separate file?

The parts between lines 76 and 158?  Sure. 

I have a couple more PRs to open, and I'll work on that next.

**Comment by kparzysz** [comment]
Tests are still coming...

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (ergawy) [inline llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:65]: I think this should work and might be a bit easier to read:
```suggestion
  std::count_if(container.begin(), container.end(), pred) == 1 ? first : container.end();
```

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (ergawy) [inline llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:76]: Can you add docs explaining the purpose of the struct and also the meaning of the template arguments? 🙏

I assume you are passing `HelperType` to share `ConstructDecompositionT` between flang and clang, right?

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (ergawy) [inline llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:107]: I think early exit here will be more consistent with [the style guide](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code).
```suggestion
      if (!success) {
        return;
      }
      // Copy the broken down directives with their clauses to the
      // output list. Copy by value, since we don't own the storage
      // with the input clauses, and the internal representation uses
      // clause addresses.
      for (auto &leaf : leafs) {
        output.push_back({leaf.id});
        auto &dwc = output.back();
        for (const ClauseTy *c : leaf.clauses)
          dwc.clauses.push_back(*c);
      }
```

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:550]: Done

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (kparzysz) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:65]: This will go over `[begin, first]` for the second time.

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (kparzysz) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:107]: Done.

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (kparzysz) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:76]: Done.

Yes, this is a class that does language-specific things that this code can't do on its own.

File: llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h
Comment (skatrak) [inline llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h:34]: Does this not introduce a significant performance penalty when all hashes in the unordered set will be guaranteed to collide?

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:2173]: Is there a reason why this can't be moved into `genOMPDispatch` as well? We won't find compound constructs here, but being consistent should simplify the reasoning about how lowering is done.

File: llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h
Comment (skatrak) [inline llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h:93]: The fundamental issue I have with this approach is that we're splitting clauses by leaf construct first and then re-combining a subset of them if it refers to a composite construct in `buildConstructQueue`. Later on, this re-combined single list of clauses will be passed to the composite lowering `genCompositeXyz` functions, which in turn will now have to re-split them so that the right clauses are attached to each leaf operation.

Wouldn't it make more sense to split once and not try to re-combine composite constructs? In particular, I'm thinking about not keeping composite constructs as a unit as a result of `buildConstructQueue` (removing the need for this `ConstructCompositionT` class) and instead updating `genOMPDispatch` to potentially match multiple leaves in the construct queue when those refer to a composite construct. Then, the single corresponding composite construct lowering function can be called passing a separate clause list for each leaf construct, removing the need to run any logic to re-attach clauses to the applicable operation again.

File: flang/test/Lower/OpenMP/copyprivate.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/copyprivate.f90:63]: It looks like most of these changes is reversing the original order of the clauses. Are these object lists already reversed in the output of `buildConstructQueue`? Or do you know where that's happening?

**Review by skatrak — COMMENTED** [review]
Thank you Krzysztof for this work, it looks very good! I have a few comments to hopefully polish things a bit before landing.

File: llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h
Comment (skatrak) [reply llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h:93]: Actually, after looking at your follow-up PR, that last part about passing multiple clause lists to `genCompositeXyz` functions wouldn't be necessary. They can just get them from the construct queue by looking at N items there rather than just one.

File: flang/test/Lower/OpenMP/copyprivate.f90
Comment (kparzysz) [reply flang/test/Lower/OpenMP/copyprivate.f90:63]: It's getting reordered in re-composition.  The DSA clauses get collapsed, i.e. `private(x) private(y)` is replaced with `private(x, y)`, or `private(y, x)`, the order there is the result of how the implementation handles the clauses---there is no predefined order.

Maybe we should sort object lists in clauses that have them...

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1434]: For `parallel sections` we process the `parallel` first with the clauses that are associated with it.  Unless we need to process the clauses from `sections` in there as well, this should work.

If we need to process extra clauses with `parallel`, it is something we will need to address elsewhere.  We should get the same behavior with `parallel sections` in a single construct, and `parallel`, `sections` written separately. The spec does not treat either of these cases differently from the other, so to get the clauses processed in the same way we'd need to have an implementation that doesn't differentiate between these two cases.

File: flang/lib/Lower/OpenMP/Decomposer.h
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/Decomposer.h]: I added a comment explaining what this function does.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2173]: I left it out because I expected a comment from you about why use the dispatch if these are simple constructs...  lol

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1967]: Added.

File: flang/test/Lower/OpenMP/copyprivate.f90
Comment (kparzysz) [reply flang/test/Lower/OpenMP/copyprivate.f90]: I changed the recomposition code to try harder to maintain predictable order, and the number of modified tests is down to 3.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1434]: Yes, it's a good point. What we had before would treat the combined form of the construct differently from nesting one construct inside of the other, when both should ideally produce the same MLIR representation.

This change makes the combined variant work the same as the nested variant and not the other way around (which I agree would be quite a bit more difficult to implement). So my concern is mainly about the chance that this might be breaking something that used to work if the right way to handle this case (according to the spec) was what was implemented for the combined variant of the construct rather than what's done for the nested variant.

By looking at the original code comments in `genOMP` it seems like this was intentional. There are a couple of tests for this in flang/test/Lower/OpenMP/parallel-sections.f90 and flang/test/Lower/OpenMP/copyin.f90. so assuming those properly test the expected behavior I suppose if they aren't broken by this change we're fine.

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (skatrak) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:53]: If you plan to refactor these, then I think it's fine to make the improvement as part of a follow-up PR.

File: llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h
Comment (skatrak) [reply llvm/include/llvm/Frontend/OpenMP/ConstructDecompositionT.h:121]: I can see where you're coming from, I guess it's largely a matter of taste since direct uses for this class will be rather restricted. I still prefer the flexibility of having a separate call to `split` because it gives us the chance to return an error code if something fails, which we would have to implement as even more state if following the constructor-only approach. And I think generally constructors aren't expected to be super-expensive to run.

Anyway, those are my two cents, but it's not a blocker for me. If you're not convinced, I think it's fine to leave that unchanged and perhaps reconsider if in the future it starts getting in the way.

File: llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h
Comment (skatrak) [reply llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h]: Thanks for the explanation, I see. I guess this is functionally kind of equivalent to creating an `std::list` and then accessing it through a set-like interface.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2175]: That's funny, you mispredicted a branch there! I think it's fine to make a follow-up PR for that.

File: llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h
Comment (skatrak) [reply llvm/include/llvm/Frontend/OpenMP/ConstructCompositionT.h:88]: Ok, we can defer these changes to another PR as well, since this already works and it's a kind of large chunk of work to move to a different approach.

**Comment by kparzysz** [comment]
I think all the points made so far were either addressed, or they will be addressed in upcoming PRs.

Are there still any remaining issues?

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1434]: I tracked down the initial addition of this handling of PARALLEL SECTIONS to patch [D122464](https://reviews.llvm.org/D122464). Maybe the original author @shraiysh or reviewers @kiranchandramohan and @NimishMishra can let us know if this change in behavior (processing clauses for SECTIONS inside of PARALLEL rather than outside) is fine. This is the only thing I'm holding off approval for, since I'm not quite sure if it's correct.

Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1434]: The change looks OK to me. Since we did not have a division of clauses before, it was probably convenient to process it all before the OpenMP ops are created.

Is this different for Parallel Do?

Comment (skatrak) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1434]: Thanks @kiranchandramohan for checking. It seems like this change should be fine then, since this was the only place something like this was done. "Parallel do" was already handled just like what this patch will do for "parallel sections".

**Review by skatrak — APPROVED** [review]
Thank you Krzysztof for working on this and addressing my comments. LGTM.

**Comment by omjavaid** [comment]
This appears to cause a regression in LLVM testsuite gfortran testsuire  FAIL: test-suite::gfortran-regression-compile-regression__gomp__reduction7_f90.test
https://lab.llvm.org/buildbot/#/builders/197/builds/14288

**Comment by kparzysz** [comment]
Thanks, I'll take a look.

**Comment by tblah** [comment]
A few tests fail for me in my local build since this commit
```
Lower/OpenMP/parallel-sections.f90
Lower/OpenMP/taskgroup.f90
Lower/OpenMP/teams.f90
```

But I see the aarch64 buildbot is green: https://lab.llvm.org/buildbot/#/builders/173 so perhaps this is a problem on my end. Do you have any idea where I should look?

Or has anyone else come across the same problem?

**Comment by kparzysz** [comment]
How does it fail?  Is it just different output, or are you seeing a crash?

**Comment by tblah** [comment]
> How does it fail? Is it just different output, or are you seeing a crash?
All three hit assertion failures in Decomposer.cpp:96: "Construct decomposition failed". This looks the same as @psteinfeld's issue.

> Edit: I'll take a look at it to see what could potentially be cause of the problem. Some earlier version of the code had non-determinism in it, maybe I haven't fixed all of that...

Thank you

**Comment by kparzysz** [comment]
> This appears to cause a regression in LLVM testsuite gfortran testsuire FAIL: test-suite::gfortran-regression-compile-regression__gomp__reduction7_f90.test https://lab.llvm.org/buildbot/#/builders/197/builds/14288

That test has a `simd` directive with a `reduction` clause that has `task` modifier, which is not valid when used with `simd`.  The purpose of the test is to detect the use of the invalid modifier.  The decomposition code was trying to apply the modifier to the construct for which it is valid (as prescribed by the spec), but in this case it hasn't found such a construct, and ended up not applying the modifier at all.  This resulted in a `reduction` clause without the modifier, which is actually valid for `simd`, and was ultimately compiled successfully, against the expectation of the test.
I put out https://github.com/llvm/llvm-project/pull/92160 for this, but the proper fix would be to detect it in the semantic checks.  The decomposition code assumes that the code it deals with is valid, and the best it can do is to assert.

**Comment by kparzysz** [comment]
> A few tests fail for me in my local build since this commit
> 
> ```
> Lower/OpenMP/parallel-sections.f90
> Lower/OpenMP/taskgroup.f90
> Lower/OpenMP/teams.f90
> ```
> 
> But I see the aarch64 buildbot is green: https://lab.llvm.org/buildbot/#/builders/173 so perhaps this is a problem on my end. Do you have any idea where I should look?
> 
> Or has anyone else come across the same problem?

I was just able to reproduce this now.  I don't know what's causing the assertion yet, but I know why it didn't fail `check-flang` for me earlier on.  The test is predicated on the presence of `omp_runtime` in the lit environment (because the test uses the `omp_lib` module).  That module wasn't built in my environment, and the test was silently skipped.  When I ran it by hand, flang-new complained about some other of its own modules not being valid.  I had to rebuild/reinstall the entire project from scratch to force these modules to be rebuilt with a recent flang-new.  FYI, the module errors I was seeing were
```
  use,intrinsic::iso_c_binding,only:c_new_line
                 ^^^^^^^^^^^^^
/work/kparzysz/c/org/bin/../include/flang/omp_lib_kinds.mod:52:16: error: Cannot read module file for module 'iso_c_binding': File is not the right module file for 'iso_c_binding': /work/kparzysz/c/org/bin/../include/flang/iso_c_binding.mod
```

**Comment by kparzysz** [comment]
> ```
> taskgroup.f90
> ```
This testcase is invalid: the `allocate` clause requires another clause that can privatize the given object.  The only such clause is `task_reduction`, but it's not yet supported.  The only way to fix the test now is to remove the `allocate` clause completely (https://github.com/llvm/llvm-project/pull/92173). My code also needs to recognize additional privatizing clauses (https://github.com/llvm/llvm-project/pull/92176).

**Comment by kparzysz** [comment]
> ```
> teams.f90
> ```
There is an `if` clause on the `teams` directive: this is only allowed in OpenMP 5.2 or later (https://github.com/llvm/llvm-project/pull/92180).

**Comment by kparzysz** [comment]
> ```
> parallel-sections.f90
> ```
This test uses `allocate` clause without a privatizing clause.  The failure is of the same type to the one in "taskgroup.f90".  One way to fix this is to add `private(x, y)` to the construct.  This would cause the `allocate` to only apply to `sections` (since that's where the `private` clause would apply).  https://github.com/llvm/llvm-project/pull/92185

**Comment by kparzysz** [comment]
> I'll work on getting a small reproducer.

@psteinfeld Could you check if adding `-fopenmp-version=52` changes anything?

**Comment by psteinfeld** [comment]
> -fopenmp-version=52

From what I can tell, this doesn't make a difference.

**Comment by tblah** [comment]
Thank you very much for the quick fixes. I'm concerned that invalid openmp directives led to assertion failures rather than semantic failures. Please could you make an issue for this so we don't forget to address it.

**Pull request merged**
